src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -506,33 +528,37 @@ uct_cuda_copy_detect_vmm(const void *address, CUmemorytype *cuda_mem_type,
         goto mem_release;
     }
 
-    /* if location type is HOST the memory can be potentially accessed by CEs */
     if ((prop.location.type == CU_MEM_LOCATION_TYPE_HOST) ||
         (prop.location.type == CU_MEM_LOCATION_TYPE_HOST_NUMA) ||
         (prop.location.type == CU_MEM_LOCATION_TYPE_HOST_NUMA_CURRENT)) {
-        *cuda_mem_type = CU_MEMORYTYPE_DEVICE;
+        *alloc_type = UCT_CUDA_ALLOC_METHOD_FABRIC_HOST;
+    } else if (prop.location.type == CU_MEM_LOCATION_TYPE_DEVICE) {
+        *alloc_type = UCT_CUDA_ALLOC_METHOD_FABRIC_DEVICE;
     } else {
+        *alloc_type = UCT_CUDA_ALLOC_METHOD_LAST;
         status = UCS_ERR_INVALID_ADDR;
     }
 
 mem_release:
     status = UCT_CUDADRV_FUNC_LOG_DEBUG(cuMemRelease(alloc_handle));
-ret:
+out:
 #endif
     return status;
 }
 
 static void
-uct_cuda_copy_set_sync_memops(const void *address, unsigned is_vmm_allocation)
+uct_cuda_copy_set_sync_memops(const void *address, CUcontext cuda_mem_ctx,
+                              unsigned is_vmm)
 {
     unsigned value = 1;
+    unsigned flags;
 
-#if HAVE_CUDA_FABRIC
-    if (is_vmm_allocation) {
-        UCT_CUDADRV_FUNC_LOG_WARN(cuCtxSetFlags(CU_CTX_SYNC_MEMOPS));
-    } else
-#endif
-    {
+    if (is_vmm) {
+        UCT_CUDADRV_FUNC_LOG_WARN(cuCtxGetFlags(&flags));
+        if (!(flags | CU_CTX_SYNC_MEMOPS)) {
+            UCT_CUDADRV_FUNC_LOG_WARN(cuCtxSetFlags(CU_CTX_SYNC_MEMOPS));
```

Comment:
do we want to affect all future allocations in this context including user allocations?
if yes what happens if we get user pointer that was allocated prior to setting this flag?